### DISCUSSION TITLE
Question regarding Spotify version 1.0.80.474 compatibility

### DISCUSSION THREAD

**Discussion by berriesdingler** [SpotX-Official/SpotX#402] [Question]
**TL;DR** 
- Is there any way for me to get SpotX working with version 1.0.80.474? Is there a log I can check for either **Rollback-Spotify** or **SpotX**, to see what commands are being causing the issue? 
- If version 1.0.80.474 is just not compatible with **SpotX**, is there any chance for the old song radio/playlist radio to be added as an experimental feature?

---

I've been using the old method of blocking Spotify ads:

1. Download Spotify 
(in my case 1.0.80.474 because I wanted to keep the old "song/playlist radio" as opposed to the "song radio" playlist in newer Spotify)
2. Create **Update** folder in **Appdata\Local\Spotify** and edit properties to read-only
7. Create and rename new .txt files to **Spotify_new.exe** and **Spotify_new.exe.sig** in **Appdata\Roaming\Spotify** and edit properties to read-only
8. Edit **Host** file  in **Windows\System32\drivers\etc** <details><summary>Host file for Spotify AdBlock</summary>
     ```
     --Spotify Ad Block
     127.0.0.1 media-match.com
     127.0.0.1 adclick.g.doublecklick.net
     127.0.0.1 googleads.g.doubleclick.net
     127.0.0.1 http://www.googleadservices...
     127.0.0.1 pagead2.googlesyndication.com
     127.0.0.1 pubads.g.doubleclick.net
     --Spotify Ads Block End
     0.0.0.0 pubads.g.doubleclick.net
     0.0.0.0 securepubads.g.doubleclick.net
     0.0.0.0 www.googletagservices.com
     0.0.0.0 gads.pubmatic.com
     0.0.0.0 ads.pubmatic.com
     0.0.0.0 tpc.googlesyndication.com
     0.0.0.0 pagead2.googlesyndication.com
     0.0.0.0 googleads.g.doubleclick.net
     0.0.0.0 adclick.g.doublecklick.net
     0.0.0.0 adeventtracker.spotify.com
     0.0.0.0 ads-fa.spotify.com
     0.0.0.0 analytics.spotify.com
     0.0.0.0 audio2.spotify.com
     0.0.0.0 b.scorecardresearch.com
     0.0.0.0 bounceexchange.com
     0.0.0.0 bs.serving-sys.com
     0.0.0.0 content.bitsontherun.com
     0.0.0.0 core.insightexpressai.com
     0.0.0.0 crashdump.spotify.com
     0.0.0.0 d2gi7ultltnc2u.cloudfront.net
     0.0.0.0 d3rt1990lpmkn.cloudfront.net
     0.0.0.0 desktop.spotify.com
     0.0.0.0 doubleclick.net
     0.0.0.0 ds.serving-sys.com
     0.0.0.0 googleadservices.com
     0.0.0.0 googleads.g.doubleclick.net
     0.0.0.0 gtssl2-ocsp.geotrust.com
     0.0.0.0 js.moatads.com
     0.0.0.0 log.spotify.com
     0.0.0.0 media-match.com
     0.0.0.0 omaze.com
     0.0.0.0 open.spotify.com
     0.0.0.0 pagead46.l.doubleclick.net
     0.0.0.0 pagead2.googlesyndication.com
     0.0.0.0 partner.googleadservices.com
     0.0.0.0 pubads.g.doubleclick.net
     0.0.0.0 redirector.gvt1.com
     0.0.0.0 s0.2mdn.net
     0.0.0.0 securepubads.g.doubleclick.net
     0.0.0.0 spclient.wg.spotify.com
     0.0.0.0 tpc.googlesyndication.com
     0.0.0.0 v.jwpcdn.com
     0.0.0.0 video-ad-stats.googlesyndication.com
     0.0.0.0 weblb-wg.gslb.spotify.com
     0.0.0.0 www.googleadservices.com
     0.0.0.0 www.googletagservices.com
     0.0.0.0 www.omaze.com
Ever since I started using this back ~2018, I've never had any problems with this setup. But last night when I tried to use Spotify, the app would continuously launch and close itself. 

I tried toubleshooting:
- Restarting my computer - Did not fix issue  
- Uninstalling and then reinstalling - This looked promising, since the app stayed open, but Spotify would automatically update to latest version. And if I repeat my old Spotify ad block setup, the endless closing and opening would start again. 


I wasn't sure if this method was patched or not so I googled/reddited (if that's a word) and found amd64fox's **Rollback-Spotify** and **SpotX**🎉🥳🎊.

- **Rollback-Spotify** did not work for me, I kept getting "Failed to block updates". I tried looking at the code to figure out why, and my best guess is that the `$old -match "(?<=wg:\/\/desktop-update\/)2(\/update)"` condition is returning false? 
  - I tried googling "(?<=wg:\/\/desktop-update\/)2(\/update)" and couldn't find anything regarding this., but I would love to learn more about this. What does `"(?<=wg:\/\/desktop-update\/.)7(\/update)"` mean/represent? and how is it different from `"(?<=wg:\/\/desktop-update\/.)2(\/update)"`?
 - Also I don't know if it's just a version  1.0.80.474 problem, but when I do a fresh install of Spotify, I can't find the **Spotify.bak** file. I don't know if this file was supposed to be part of the installion or if **Rollback-Spotify** creates such a file (like if I reference a non-existent folder when extracting a .zip, the folder is freshly created), so I tried running **Rollback-Spotify in admin mode. But that did not work either. 
- Since **Rollback-Spotify** did not work for me, I decided to try **SpotX**. But for my initial setup configuration:
   ```Your Spotify 1.0.80.474 version is outdated, it is reecommended to upgrade to 1.2.10.760
   Want to update? [Y/N]: n

   Do you want to disable podcasts, episodes and audiobooks from the main page? [Y/N]: y

   Want to block Spotify updates? [Y/N]
   It is recommended to block because Spotify was downgraded: y

   Do you want to install automatic audio cache cleaning for Spotify? [Y/N]: n

   Patching Spotify...

   xpui.spa not found, reinstall Spotify
   Script is stopped
   Press Enter to continue...
   ```
  - And like with **Spotify.bak**, I'm not sure if this is just a me issue or if version 1.0.80.474 just doesn't have the xpui.spa file upon default installation.

- Finally, I tried upgrading to version 1.2.10.760, per SpotX recommendation, and this ended up working for me.

---

Apologies for the long preamble, my questions are:
1. Is there any way for me to get SpotX working with version 1.0.80.474? Is there a log I can check for either **Rollback-Spotify** or **SpotX**, to see what commands are being causing the issue? 
2. If version 1.0.80.474 is just not compatible with **SpotX**, is there any chance for the old song radio/playlist radio to be added as an experimental feature?

---
---

One more question, not specific to version 1.80.474 but in general, is there any way for me to make it so that Spotify automatically starts in ***Private Session***, similar to how login info can be stored in the **pref** file **Appdata\Roaming\Spotify**

**Comment by amd64fox** [comment]
> 1. Is there any way for me to get SpotX working with version 1.0.80.474? Is there a log I can check for either **Rollback-Spotify** or **SpotX**, to see what commands are being causing the issue?

Spotx and Rollback-Spotify are naturally not compatible with such an old version as 1.0.80.474, if I'm not mistaken this version is about 5 years old, and SpotX was created about a year ago.

The problem is that your Spotify 1.0.80.474 started updating itself forcibly, is that correct?

> 2\. is there any chance for the old song radio/playlist radio to be added as an experimental feature?

question to Spotify developers

> is there any way for me to make it so that Spotify automatically starts in _**Private Session**_

Theoretically, this is most likely possible, but how everything will be in practice, until you take it, will not know

**Reply by berriesdingler** [reply]
> The problem is that your Spotify 1.0.80.474 started updating itself forcibly, is that correct?

Yeah, the 1.0.80.474 version automatically updates by default as soon as I enter login credentials. If I apply the adblocking procedure that I used to use, the app would just endlessly cycle between opening and closing. 

For now, I'm down to give the ver 1.2 a try, from the past day of use I've noticed small inconvenience such as:
- In order to get to the equivalent of the old song/playlist radio, I need to go to **playlist radio** of the **song radio**.
- The **recently played** section on either the **Home** or the tab from the **Queue** do not display recent **playlist radio**. 
But these aren't really dealbreakers.


---

On another note, would you happen to any resources you would recommend for me if I want to learn more about Spotify adblocking and changing/adding features?

I'm just starting, and I've been trying to look at the availble Github codes on projects such as SpotX and Spicetify. But I understand the bare bare minimum of whats going on. While I try to google what I don't understand, sometimes I still can't find answers. 

Maybe I'm just not a great googler, because I see snipits of the code like `"(?<=wg:\/\/desktop-update\/.)7(\/update)"` or 
```
            "match": [
                "}\\[dir=ltr\\]\\s?([.a-zA-Z\\d[_]+?,\\[dir=ltr\\])",
                "}\\[dir=ltr\\]\\s?",
                "html\\[dir=ltr\\]",
                ",\\s?\\[dir=rtl\\].+?(\\{.+?\\})",
                "[\\w\\-\\.]+\\[dir=rtl\\].+?\\{.+?\\}",
                "\\}\\[lang=ar\\].+?\\{.+?\\}",
                "\\}\\[dir=rtl\\].+?\\{.+?\\}",
                "\\}html\\[dir=rtl\\].+?\\{.+?\\}",
                "\\}html\\[lang=ar\\].+?\\{.+?\\}",
                "\\[lang=ar\\].+?\\{.+?\\}",
                "html\\[dir=rtl\\].+?\\{.+?\\}",
                "html\\[lang=ar\\].+?\\{.+?\\}",
                "\\[dir=rtl\\].+?\\{.+?\\}",
                "\\[dir=str\\]",
                "[/]\\*([^*]|[\r\n]|(\\*([^/]|[\r\n])))*\\*[/]",
                "[/][/]#\\s.*",
                "\r?\n(?!\\(1|\\d)"
            ],
```

and I can't find anything when googling. It's not the syntax that I have trouble understanding (that I can google, and their deifnition/uses seems pretty ubiquitous), but it's more about how do you even know what you're doing/looking at?

**Reply by amd64fox** [reply]
> If I apply the adblocking procedure that I used to use, the app would just endlessly cycle between opening and closing.

To block the update to 1.0.80.474, you can try another method:

- Install version 1.0.80.474 and immediately close the application
- Go to the `AppData\Local\Spotify` folder
- Create a file without an extension named `Update`
- In the properties of this file, specify the `read-only` attribute

Now, log in to the client and check if it will update automatically.
